File: backend/src/models/Producto.js
const { DataTypes } = require('sequelize');
const sequelize = require('../config/database');

const Producto = sequelize.define('Producto', {
  id: {
    type: DataTypes.UUID,
    defaultValue: DataTypes.UUIDV4,
    primaryKey: true,
  },
  emprendimientoId: {
    type: DataTypes.UUID,
    allowNull: false,
    field: 'emprendimiento_id',
  },
  recetaId: {
    type: DataTypes.UUID,
    allowNull: true,
    field: 'receta_id',
    comment: 'Receta que define el costo de producción del producto (opcional)',
  },
  nombre: {
    type: DataTypes.STRING(150),
    allowNull: false,
  },
  categoria: {
    type: DataTypes.STRING(100),
  },
  descripcion: {
    type: DataTypes.TEXT,
  },
  imagenUrl: {
    type: DataTypes.STRING(255),
    field: 'imagen_url',
  },
  costoUnitario: {
    type: DataTypes.DECIMAL(12, 2),
    allowNull: false,
    defaultValue: 0,
    field: 'costo_unitario',
  },
  margenGanancia: {
    type: DataTypes.DECIMAL(5, 2),
    allowNull: false,
    defaultValue: 40,
    field: 'margen_ganancia',
  },
  precioVenta: {
    type: DataTypes.DECIMAL(12, 2),
    allowNull: false,
    defaultValue: 0,
    field: 'precio_venta',
  },
  controlaStock: {
    type: DataTypes.BOOLEAN,
    defaultValue: true,
    field: 'controla_stock',
  },
  stockActual: {
    type: DataTypes.DECIMAL(12, 2),
    allowNull: false,
    defaultValue: 0,
    field: 'stock_actual',
  },
  stockMinimo: {
    type: DataTypes.DECIMAL(12, 2),
    allowNull: false,
    defaultValue: 0,
    field: 'stock_minimo',
  },
  activo: {
    type: DataTypes.BOOLEAN,
    defaultValue: true,
  },
}, {
  tableName: 'productos',
});

module.exports = Producto;

File: mobile/lib/models/producto.dart
import 'parsing.dart';

class Producto {
  final String id;
  final String nombre;
  final String? categoria;
  final String? descripcion;
  final String? recetaId;
  final double costoUnitario;
  final double margenGanancia;
  final double precioVenta;
  final bool controlaStock;
  final double stockActual;
  final double stockMinimo;
  final bool activo;

  Producto({
    required this.id,
    required this.nombre,
    this.categoria,
    this.descripcion,
    this.recetaId,
    required this.costoUnitario,
    required this.margenGanancia,
    required this.precioVenta,
    required this.controlaStock,
    required this.stockActual,
    required this.stockMinimo,
    required this.activo,
  });

  bool get bajoStock => controlaStock && stockActual <= stockMinimo;

  double get gananciaUnitaria => precioVenta - costoUnitario;

  factory Producto.fromJson(Map<String, dynamic> json) {
    return Producto(
      id: json['id'] as String,
      nombre: json['nombre'] as String,
      categoria: json['categoria'] as String?,
      descripcion: json['descripcion'] as String?,
      recetaId: json['recetaId'] as String?,
      costoUnitario: parseDouble(json['costoUnitario']),
      margenGanancia: parseDouble(json['margenGanancia']),
      precioVenta: parseDouble(json['precioVenta']),
      controlaStock: json['controlaStock'] as bool? ?? true,
      stockActual: parseDouble(json['stockActual']),
      stockMinimo: parseDouble(json['stockMinimo']),
      activo: json['activo'] as bool? ?? true,
    );
  }

  Map<String, dynamic> toJson() => {
        'nombre': nombre,
        'categoria': categoria,
        'descripcion': descripcion,
        'recetaId': recetaId,
        'costoUnitario': costoUnitario,
        'margenGanancia': margenGanancia,
        'precioVenta': precioVenta,
        'controlaStock': controlaStock,
        'stockActual': stockActual,
        'stockMinimo': stockMinimo,
      };
}
